File: AIAgent/ZBuild/Repos/cxdb-main/clients/go/fstree/options.go
// Copyright 2025 StrongDM Inc
// SPDX-License-Identifier: Apache-2.0

package fstree

import "path/filepath"

// Option configures snapshot behavior.
type Option func(*options)

type options struct {
	excludePatterns []string
	excludeFn       func(path string, isDir bool) bool
	followSymlinks  bool
	maxFileSize     int64
	maxFiles        int
}

func defaultOptions() *options {
	return &options{
		excludePatterns: nil,
		followSymlinks:  false,
		maxFileSize:     100 * 1024 * 1024, // 100MB default max file size
		maxFiles:        100000,            // 100k files max
	}
}

// WithExclude adds glob patterns for paths to exclude.
// Patterns are matched against the relative path from the root.
// Examples: "*.log", ".git/**", "node_modules/**"
func WithExclude(patterns ...string) Option {
	return func(o *options) {
		o.excludePatterns = append(o.excludePatterns, patterns...)
	}
}

// WithExcludeFunc sets a custom exclusion function.
// Return true to exclude the path. Called for every file and directory.
func WithExcludeFunc(fn func(path string, isDir bool) bool) Option {
	return func(o *options) {
		o.excludeFn = fn
	}
}

// WithFollowSymlinks enables following symbolic links.
// By default, symlinks are captured as symlinks (their target path is stored).
// With this option, symlinks are dereferenced and their target content is captured.
// Circular symlinks are detected and skipped.
func WithFollowSymlinks() Option {
	return func(o *options) {
		o.followSymlinks = true
	}
}

// WithMaxFileSize sets the maximum file size to include.
// Files larger than this are skipped. Default is 100MB.
func WithMaxFileSize(bytes int64) Option {
	return func(o *options) {
		o.maxFileSize = bytes
	}
}

// WithMaxFiles sets the maximum number of files to include.
// Default is 100,000.
func WithMaxFiles(n int) Option {
	return func(o *options) {
		o.maxFiles = n
	}
}

// shouldExclude checks if a path should be excluded based on options.
func (o *options) shouldExclude(relPath string, isDir bool) bool {
	// Check custom function first
	if o.excludeFn != nil && o.excludeFn(relPath, isDir) {
		return true
	}

	// Check glob patterns
	for _, pattern := range o.excludePatterns {
		// Try direct match
		if matched, _ := filepath.Match(pattern, relPath); matched {
			return true
		}
		// Try matching just the base name
		if matched, _ := filepath.Match(pattern, filepath.Base(relPath)); matched {
			return true
		}
		// For ** patterns, do prefix matching on directories
		if isDir && len(pattern) > 3 && pattern[len(pattern)-3:] == "/**" {
			prefix := pattern[:len(pattern)-3]
			if matched, _ := filepath.Match(prefix, relPath); matched {
				return true
			}
		}
	}

	return false
}

File: AIAgent/ZBuild/Repos/cxdb-main/clients/rust/src/fstree/options.rs
// Copyright 2025 StrongDM Inc
// SPDX-License-Identifier: Apache-2.0

#![allow(clippy::type_complexity)]

use std::path::Path;
use std::sync::Arc;

use glob::Pattern;

pub type SnapshotOption = Arc<dyn Fn(&mut Options) + Send + Sync>;

#[derive(Clone)]
pub struct Options {
    pub exclude_patterns: Vec<String>,
    pub exclude_fn: std::option::Option<Arc<dyn Fn(&str, bool) -> bool + Send + Sync>>,
    pub follow_symlinks: bool,
    pub max_file_size: i64,
    pub max_files: usize,
}

impl Default for Options {
    fn default() -> Self {
        Self {
            exclude_patterns: Vec::new(),
            exclude_fn: None,
            follow_symlinks: false,
            max_file_size: 100 * 1024 * 1024,
            max_files: 100_000,
        }
    }
}

pub fn with_exclude(patterns: impl IntoIterator<Item = impl Into<String>>) -> SnapshotOption {
    let patterns: Vec<String> = patterns.into_iter().map(|p| p.into()).collect();
    Arc::new(move |opts| {
        opts.exclude_patterns.extend(patterns.clone());
    })
}

pub fn with_exclude_func<F>(func: F) -> SnapshotOption
where
    F: Fn(&str, bool) -> bool + Send + Sync + 'static,
{
    let func = Arc::new(func);
    Arc::new(move |opts| opts.exclude_fn = Some(func.clone()))
}

pub fn with_follow_symlinks() -> SnapshotOption {
    Arc::new(|opts| opts.follow_symlinks = true)
}

pub fn with_max_file_size(bytes: i64) -> SnapshotOption {
    Arc::new(move |opts| opts.max_file_size = bytes)
}

pub fn with_max_files(count: usize) -> SnapshotOption {
    Arc::new(move |opts| opts.max_files = count)
}

impl Options {
    pub fn should_exclude(&self, rel_path: &str, is_dir: bool) -> bool {
        if let Some(func) = &self.exclude_fn {
            if func(rel_path, is_dir) {
                return true;
            }
        }

        let rel_path = normalize_path(rel_path);
        let basename = Path::new(&rel_path)
            .file_name()
            .and_then(|s| s.to_str())
            .unwrap_or("");

        for pattern in &self.exclude_patterns {
            if is_double_star_dir(pattern, &rel_path, is_dir) {
                return true;
            }
            if matches_glob(pattern, &rel_path) {
                return true;
            }
            if matches_glob(pattern, basename) {
                return true;
            }
        }
        false
    }
}

fn normalize_path(path: &str) -> String {
    path.replace('\\', "/")
}

fn matches_glob(pattern: &str, path: &str) -> bool {
    Pattern::new(pattern)
        .map(|p| p.matches(path))
        .unwrap_or(false)
}

fn is_double_star_dir(pattern: &str, rel_path: &str, is_dir: bool) -> bool {
    if !is_dir {
        return false;
    }
    if let Some(prefix) = pattern.strip_suffix("/**") {
        if rel_path == prefix {
            return true;
        }
        let prefix_with_sep = format!("{prefix}/");
        return rel_path.starts_with(&prefix_with_sep) || matches_glob(prefix, rel_path);
    }
    false
}
